Extensions/Share/ShareViewController.swift

```diff
@@ -0,0 +1,303 @@
+        let row = UIView()
+
+        let stackView = UIStackView()
+        stackView.axis = .vertical
+        stackView.spacing = UX.pageInfoLineSpacing
+
+        row.addSubview(stackView)
+        stackView.snp.makeConstraints { make in
+            make.centerY.equalToSuperview()
+            make.left.right.equalToSuperview().inset(UX.pageInfoRowLeftInset)
+        }
+
+        let pageTitleLabel = UILabel()
+        let urlLabel = UILabel()
+        [pageTitleLabel, urlLabel].forEach { label in
+            stackView.addArrangedSubview(label)
+            label.allowsDefaultTighteningForTruncation = true
+            label.lineBreakMode = .byTruncatingMiddle
+            label.font = UX.baseFont
+        }
+
+        pageTitleLabel.font = UIFont.boldSystemFont(ofSize: UX.baseFont.pointSize)
+
+        return (row, pageTitleLabel, urlLabel)
+    }
+
+    private func makeActionRow(label: String, imageName: String, action: Selector, hasNavigation: Bool) -> UIView {
+        let row = UIView()
+
+        let stackView = UIStackView()
+        stackView.axis = .horizontal
+        stackView.spacing = UX.actionRowSpacingBetweenIconAndTitle
+        row.addSubview(stackView)
+
+        stackView.snp.makeConstraints { make in
+            make.top.bottom.equalToSuperview()
+            make.left.right.equalToSuperview().inset(UX.pageInfoRowLeftInset)
+        }
+
+        let icon = UIImageView(image: UIImage(named: imageName))
+        icon.contentMode = .scaleAspectFit
+
+        let title = UILabel()
+        title.font = UX.baseFont
+        title.numberOfLines = 2
+        title.adjustsFontSizeToFitWidth = true
+        title.allowsDefaultTighteningForTruncation = true
+
+        title.text = label
+        [icon, title].forEach { stackView.addArrangedSubview($0) }
+
+        icon.snp.makeConstraints { make in
+            make.size.equalTo(UX.actionRowIconSize)
+        }
+
+        if hasNavigation {
+            let navButton = UIImageView(image: UIImage(named: "menu-Disclosure"))
+            navButton.contentMode = .scaleAspectFit
+            stackView.addArrangedSubview(navButton)
+            navButton.snp.makeConstraints { make in
+                make.size.equalTo(14)
+            }
+        }
+
+        let gesture = UITapGestureRecognizer(target: self, action: action)
+        row.addGestureRecognizer(gesture)
+
+        actionRows.append(row)
+        return row
+    }
+
+    fileprivate func animateToActionDoneView(withTitle title: String = "") {
+        navigationItem.leftBarButtonItem = nil
+
+        navigationController?.view.snp.updateConstraints {
+            make in
+            make.height.equalTo(UX.viewHeightForDoneState)
+        }
+
+        UIView.animate(withDuration: 0.2, animations: {
+            self.actionRows.forEach { $0.removeFromSuperview() }
+            self.separators.forEach { $0.removeFromSuperview() }
```

Comment:
It might be simpler to keep a reference to the `PageInfoRow` instead of keeping a ref to all the actions and separators. This has the added benefit of also removing the whitespace view while animating.

Reply:
The trailing whitespace is needed for the animation or the pageInfo label moves. Even setting pageInfo label constraint to `greaterthan(the height)` as it is now doesn't ensure this label doesn't move during the animation. I'll add a comment to the code to mention that.